control/discovery.py

```diff
@@ -0,0 +1,1077 @@
+            return -1
+        self.logger.debug("reply set feature request.")
+        return 0
+
+    def reply_get_feature(self, conn, data, cmd_id):
+        """Reply get feature request."""
+
+        self.logger.debug("handle get feature request.")
+        self_conn = self.conn_vals[conn.fileno()]
+        nvme_nsid = struct.unpack_from('<I', data, 12)[0]
+        nvme_rsvd1 = struct.unpack_from('<Q', data, 16)[0]
+        nvme_mptr = struct.unpack_from('<Q', data, 24)[0]
+        nvme_sgl = struct.unpack_from('<16B', data, 32)
+        nvme_sgl_desc_type = nvme_sgl[15] & 0xF0
+        nvme_sgl_desc_sub_type = nvme_sgl[15] & 0x0F
+        nvme_get_features_dword10 = struct.unpack_from('<I', data, 48)[0]
+        nvme_get_features_dword11 = struct.unpack_from('<I', data, 52)[0]
+        nvme_get_features_dword12 = struct.unpack_from('<I', data, 56)[0]
+        nvme_get_features_dword13 = struct.unpack_from('<I', data, 60)[0]
+        nvme_get_features_dword14 = struct.unpack_from('<I', data, 64)[0]
+        nvme_get_features_dword15 = struct.unpack_from('<I', data, 68)[0]
+
+        pdu_reply = Pdu()
+        pdu_reply.type = NVME_TCP_PDU.RSP
+        pdu_reply.header_length = 24
+        pdu_reply.packet_length = 24
+
+        get_feature_reply =  CqeNVMe()
+        get_feature_reply.dword0 = self_conn.keep_alive_timeout
+        get_feature_reply.sq_head_ptr = self_conn.sq_head_ptr
+        get_feature_reply.cmd_id = cmd_id
+
+        try:
+            conn.sendall(pdu_reply + get_feature_reply)
+        except BrokenPipeError:
+            self.logger.error("client disconnected unexpectedly.")
+            return -1
+        self.logger.debug("reply get feature request.")
+        return 0
+
+    def reply_get_log_page(self, conn, data, cmd_id):
+        """Reply get log page request."""
+
+        self.logger.debug("handle get log page request.")
+        self_conn = self.conn_vals[conn.fileno()]
+        my_omap_dict = self._read_all()
+        listeners = self._get_vals(my_omap_dict, self.LISTENER_PREFIX)
+        hosts = self._get_vals(my_omap_dict, self.HOST_PREFIX)
+        if len(self_conn.nvmeof_connect_data_hostnqn) != 256:
+            self.logger.error("error hostnqn.")
+            return -1
+        hostnqn = "".join(chr(byte) for byte \
+            in self_conn.nvmeof_connect_data_hostnqn[:256]).rstrip('\x00')
+
+        nvme_nsid = struct.unpack_from('<I', data, 12)[0]
+        nvme_rsvd1 = struct.unpack_from('<Q', data, 16)[0]
+        nvme_mptr = struct.unpack_from('<Q', data, 24)[0]
+        nvme_sgl = struct.unpack_from('<16B', data, 32)
+        nvme_sgl_desc_type = nvme_sgl[15] & 0xF0
+        nvme_sgl_desc_sub_type = nvme_sgl[15] & 0x0F
+        nvme_get_logpage_dword10 = struct.unpack_from('<I', data, 48)[0]
+        # nvme_get_logpage_numd indicate the reply bytes, rule: (values+1)*4
+        nvme_get_logpage_numd = struct.unpack_from('<I', data, 50)[0]
```

Comment:
16 bit integer is being read, but `I` format, which represents an unsigned integer of size 4 bytes.
Is `H` format  (unsigned 2-byte integer)  more appropriate?
```suggestion
        nvme_get_logpage_numd = struct.unpack_from('<H', data, 50)[0]
```

Reply:
No, I read 4 bytes here. This may indeed cause misunderstandings, but this field spans between `dword10` and` dword11`. Generally speaking, the `dword` field has no specific meaning, and the one I named has a specific meaning. saw  the pictire.
![image](https://github.com/ceph/ceph-nvmeof/assets/13258081/be9dedf5-e821-44a7-8fb6-68c827b96b88)